src/plugins/unified_search/public/dataview_picker/data_view_selector.tsx

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { Fragment, useEffect, useRef, useState } from 'react';
+import type { EuiSelectableProps } from '@elastic/eui';
+import type { DataViewListItem } from '@kbn/data-views-plugin/public';
+import { useKibana } from '@kbn/kibana-react-plugin/public';
+import { DataViewsList } from './dataview_list';
+import { IUnifiedSearchPluginServices } from '../types';
+import { ExploreMatchingButton } from './explore_matching_button';
+
+interface DataViewSelectorProps {
+  currentDataViewId?: string;
+  searchListInputId?: string;
+  dataViewsList: DataViewListItem[];
+  selectableProps?: EuiSelectableProps;
+  isTextBasedLangSelected: boolean;
+  setPopoverIsOpen: (isOpen: boolean) => void;
+  onChangeDataView: (dataViewId: string) => void;
+  onCreateDefaultAdHocDataView?: (pattern: string) => void;
+}
+
+export const DataViewSelector = ({
+  currentDataViewId,
+  searchListInputId,
+  dataViewsList,
+  selectableProps,
+  isTextBasedLangSelected,
+  setPopoverIsOpen,
+  onChangeDataView,
+  onCreateDefaultAdHocDataView,
+}: DataViewSelectorProps) => {
+  const kibana = useKibana<IUnifiedSearchPluginServices>();
+  const { dataViews } = kibana.services;
+
+  const [noDataViewMatches, setNoDataViewMatches] = useState(false);
+  const [dataViewSearchString, setDataViewSearchString] = useState('');
+  const [indexMatches, setIndexMatches] = useState(0);
+
+  const pendingIndexMatch = useRef<undefined | NodeJS.Timeout>();
+
+  useEffect(() => {
+    async function checkIndices() {
+      if (dataViewSearchString !== '' && noDataViewMatches) {
+        const matches = await dataViews.getIndices({
+          pattern: dataViewSearchString,
+          isRollupIndex: () => false,
+          showAllIndices: false,
+        });
+        setIndexMatches(matches.length);
+      }
+    }
+    if (pendingIndexMatch.current) {
+      clearTimeout(pendingIndexMatch.current);
+    }
+    pendingIndexMatch.current = setTimeout(checkIndices, 250);
```

Comment:
Should we clear the timeout when the component unmounts too?

Reply:
Nice catch, I was just moving this logic to a different component. Fixed this.

Reply:
[`Cleanup`](https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1) is also runned before next effect. _React also cleans up effects from the previous render before running the effects next time_

Reply:
Oh yeah.. I guess that would only make sense or effects wouldn't work 😅 Apologies, too much reviewing for one day, this is definitely working as intended now.